graylog2-server/src/main/java/org/graylog2/entitygroups/model/DBEntityGroupService.java

```diff
@@ -0,0 +1,164 @@
+
+        final IndexOptions entityTypeOptions = new IndexOptions()
+                .collation(Collation.builder().locale("en").collationStrength(CollationStrength.SECONDARY).build())
+                .unique(false)
+                .sparse(true);
+        // Add wildcard index on the entities map keys so that indices are created for any entity types that are added.
+        collection.createIndex(new BasicDBObject(EntityGroup.FIELD_ENTITIES + ".*", 1), entityTypeOptions);
+    }
+
+    public Optional<EntityGroup> get(String id) {
+        return mongoUtils.getById(id);
+    }
+
+    public PaginatedList<EntityGroup> findPaginated(String query, int page, int perPage, Bson sort, Predicate<EntityGroup> filter) {
+        final SearchQuery searchQuery = searchQueryParser.parse(query);
+        return filter == null ?
+                paginationHelper.filter(searchQuery.toBson()).sort(sort).perPage(perPage).page(page) :
+                paginationHelper.filter(searchQuery.toBson()).sort(sort).perPage(perPage).page(page, filter);
+    }
+
+    public PaginatedList<EntityGroup> findPaginatedForEntity(String type, String entityId, int page, int perPage, Bson sort) {
+        final Bson query = and(
+                exists(typeField(type)),
+                in(typeField(type), entityId)
+        );
+
+        return paginationHelper.filter(query).sort(sort).perPage(perPage).page(page);
+    }
+
+    public EntityGroup save(EntityGroup entityGroup) {
+        final String newId = scopedEntityMongoUtils.create(entityGroup);
+        return entityGroup.toBuilder().id(newId).build();
+    }
+
+    public EntityGroup update(EntityGroup entityGroup) {
+        return scopedEntityMongoUtils.update(entityGroup);
+    }
+
+    public EntityGroup addEntityToGroup(String groupId, String type, String entityId) {
+        return collection.findOneAndUpdate(MongoUtils.idEq(groupId),
+                addToSet(typeField(type), entityId),
+                new FindOneAndUpdateOptions().returnDocument(ReturnDocument.AFTER));
+    }
+
+    public Optional<EntityGroup> getByName(String name) {
+        final Bson query = eq(EntityGroup.FIELD_NAME, name);
+
+        return Optional.ofNullable(collection.find(query).first());
+    }
+
+    public Stream<EntityGroup> streamAllForEntity(String type, String entityId) {
+        final Bson query = and(
+                exists(typeField(type)),
+                in(typeField(type), entityId)
+        );
+        return MongoUtils.stream(collection.find(query));
+    }
+
+    public Multimap<String, EntityGroup> getAllForEntities(String type, Collection<String> entityIds) {
+        final Bson query = and(
+                exists(typeField(type)),
+                in(typeField(type), entityIds)
+        );
+        final Multimap<String, EntityGroup> entityToGroupsMap = MultimapBuilder.hashKeys().hashSetValues().build();
+        MongoUtils.stream(collection.find(query)).forEach(group -> {
```

Comment:
Stream uses must call close, because this is an IO stream (see the javadocs of MongoUtils#stream).
`.forEach` does not call `close()`, although one could assume it does.

The best way to do this is always to use try-with-resources on stream usages; worst case, it's a no-op for streams that don't require it:

```java
try (final Stream<EntityGroup> stream = MongoUtils.stream(collection.find(query))) {
            stream.forEach(group -> { ... });
}
```

Reply:
Wow I was not aware of this, thanks for pointing that out! I've addressed it here and will work this pattern in elsewhere as  I see it.